### ISSUE TITLE
clang-format-13 crashes with AlignArrayOfStructures

### ISSUE THREAD

**Issue by omartijn** [llvm/llvm-project#51277]
|  |  |
| --- | --- |
| Bugzilla Link | [51935](https://llvm.org/bz51935) |
| Version | unspecified |
| OS | Linux |
| Attachments | [The format file](https://user-images.githubusercontent.com/92601861/143762560-2055a8d9-2fb4-498c-ae12-2d5b5fcbf1aa.gz), [The file that fails](https://user-images.githubusercontent.com/92601861/143762561-bd71f21c-104a-4213-b7b8-8a2830a57cd8.gz) |

## Extended Description

**Comment by mkurdej** [comment]
Minimum .clang-format config file that reproduces the issue:
```
AlignArrayOfStructures: Left
```

**Comment by mkurdej** [comment]
And the minimum file:
```
void f()
{
  S{{}, {}, { a, b}};
}
```
It also crashes with a comma after `b}`:
```
void f()
{
  S{{}, {}, { a, b},};
}
```

It doesn't crash if there are less than two `{}` before `{a, b}` or if there are some after that.

**Comment by mkurdej** [comment]
Reproducers:
```
  verifyFormat("S{\n"
               "    {},\n"
               "    {},\n"
               "    {a, b}\n"
               "};\n",
               Style);
  verifyFormat("S{\n"
               "    {},\n"
               "    {},\n"
               "    {a, b},\n"
               "};\n",
               Style);
```

From what I've seen, the problem is in at least two places (probably more because I haven't checked `alignArrayInitializersRightJustified` function which has probably the same symptom).
In `WhitespaceManager::alignArrayInitializersLeftJustified`, the `Start` and `End` are computed as:
```
      auto *Start = (Cells.begin() + RowCount * CellDescs.CellCount);
      auto *End = Start + Offset;
```
but `RowCount * CellDescs.CellCount` should be `< CellDescs.Cells.size()`
and `RowCount * CellDescs.CellCount + Offset ` should be `<= CellDescs.Cells.size()`.
Which is not the case here (RowCount=3, CellCount=3, Cells has 7 elements).
The very same problem is also in `getMaximumNetWidth`.

Not sure how to fix it properly though.

**Comment by mkurdej** [comment]
BTW, the crash happens on Windows too.

**Comment by Yoshi2889** [comment]
The following snippet also crashes for me on macOS 12.1 with clang-format 14.0.0 bundled with CLion 2021.3.1:

```cxx

#include <utility>
#include <vector>

namespace stuff {
class test {
  std::vector<std::vector<int>> stuff;

public:
  test(int width, int height) {}

  explicit test(std::vector<std::vector<int>> stuff)
      : stuff(std::move(stuff)){};
};
} // namespace stuff

int main(int argc, char *argv[]) {
  int something = 10;
  auto thing = stuff::test{
      {{10, 11}, // first row
       {12, 13}, // second row

       {14, something}, //third row
       {15, 16}}
  };
}

```

The used .clang-format is:
```
BasedOnStyle: LLVM
AlignArrayOfStructures: Left
```

If I remove the whitespace between the second and third rows it works.

**Comment by StephanTLavavej** [comment]
I believe that this is the same bug, but I thought it might be helpful to provide my reduced test case (just need two layers of braces). Repros with clang-format 13 on Windows provided by VS 2022 17.1 Preview 2.

```
C:\Temp>type test.cpp
```
```cpp
struct Array {
    int elems[4];
};

int main() {
    Array a{{
        11,
        22,
        33,
        44,
    }};
}
```
```
C:\Temp>clang-format --version
clang-format version 13.0.0

C:\Temp>clang-format -style="{BasedOnStyle: LLVM}" test.cpp
struct Array {
  int elems[4];
};

int main() {
  Array a{{
      11,
      22,
      33,
      44,
  }};
}

C:\Temp>clang-format -style="{BasedOnStyle: LLVM, AlignArrayOfStructures: Left}" test.cpp
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: clang-format "-style={BasedOnStyle: LLVM, AlignArrayOfStructures: Left}" test.cpp
 #0 0x00007ff6d576ccd0 C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-format.exe 0xbccd0 C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-format.exe 0xbc719
 #1 0x00007ff6d576ccd0 C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-format.exe 0xbb2a9 C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-format.exe 0xb84e7
 #2 0x00007ff6d576ccd0 C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-format.exe 0x80b28 C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-format.exe 0x95c37
 #3 0x00007ff6d576ccd0 C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-format.exe 0x8cb14 C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-format.exe 0x77759
 #4 0x00007ff6d576ccd0 C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-format.exe 0x78019 C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-format.exe 0x44b2
 #5 0x00007ff6d576ccd0 C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-format.exe 0x2fc6 C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-format.exe 0x179c84
 #6 0x00007ff6d576ccd0 (C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-format.exe+0xbccd0)
 #7 0x00007ff6d576c719 (C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\Llvm\x64\bin\clang-format.exe+0xbc719)
0x00007FF6D576CCD0 (0x0000000000000086 0xFFFFFFFF00000021 0x0000000000000000 0x000000B97518CBB8)
0x00007FF6D576C719 (0x00000021FFFFFFFF 0x0000000000000000 0x000000B97518CBB8 0x0000000000000000)
0x00007FF6D576B2A9 (0x0000000000000000 0x000000B900000000 0x000000B97518E6E8 0x00007FFFABAD218B)
0x00007FF6D57684E7 (0x0000000000000001 0x0000000000000000 0x0000000000000081 0x0000029AC9170150)
0x00007FF6D5730B28 (0x000000B97518C220 0x0000000000000002 0x000000B97518BED0 0x00000002016DC68D)
0x00007FF6D5745C37 (0x000000B97518CFF0 0x0000000000000008 0x00007FF6D582E065 0x0000000000000002)
0x00007FF6D573CB14 (0x0000029AC9189CF0 0x0000000000000000 0x0000000000000000 0x0000000000000000)
0x00007FF6D5727759 (0x0000029AC9188770 0x00007FFFAE1E7551 0x0000029AC9188760 0x0000029AC9170000)
0x00007FF6D5728019 (0x0000000000000040 0x0000000000000000 0x0000000000000000 0x0000000000000118)
0x00007FF6D56B44B2 (0x0000000000000000 0x000000B900000001 0x0000029AC9170000 0x0000000F00000000)
0x00007FF6D56B2FC6 (0x0000000000000000 0x00007FF6D5829CFD 0x0000000000000000 0x0000000000000000)
0x00007FF6D5829C84 (0x0000000000000000 0x0000000000000000 0x0000000000000000 0x0000000000000000)
0x00007FFFAD6D54E0 (0x0000000000000000 0x0000000000000000 0x0000000000000000 0x0000000000000000), BaseThreadInitThunk() + 0x10 bytes(s)
0x00007FFFAE1C485B (0x0000000000000000 0x0000000000000000 0x0000000000000000 0x0000000000000000), RtlUserThreadStart() + 0x2B bytes(s)
```

**Comment by mkurdej** [comment]
The code:
```
struct Array {
    int elems[4];
};

int main() {
    Array a{{
        11,
        22,
        33,
        44,
    }};
}
```
does not crash on current main, but it yields misformatted code (elements grouped by two, braces not aligned):
```
struct Array {
  int elems[4];
};

int main() {
  Array a{
      {
       11, 22,
       33, 44,
       }
  };
}
```

The original poster's code still crashes.

**Comment by mydeveloperday** [comment]
https://reviews.llvm.org/D121069 fixes the majority of the producers here, but not the  `{14,something }, // third row` one, I'm looking into that as it considers the array square but runs off the end of the "Cells"

**Comment by rymiel** [comment]
The crash was fixed by adfe58b09df957f8911c5e152daa9564e5931631, and while the format isn't ideal after that patch, this issue tracks a crash which has been fixed.
